src/lug/module_detection.py

```diff
@@ -0,0 +1,130 @@
+import ast
+import glob
+import inspect
+import sys
+import types
+import os
+
+from .constants import STDLIB_MODULE_NAMES
+
+
+def parse_ast_for_file(python_file_path):
+    opened_file = open(python_file_path).read()
+    parsed_ast = ast.parse(opened_file)
+    return [node for node in ast.walk(parsed_ast)]
+
+
+def get_registered_module_imports(modules_to_register, modules_to_skip):
+    filtered_sys_modules = []
+    for module_name, module_value in sys.modules.items():
+        is_builtin_module = module_name in sys.builtin_module_names
+        is_stdlib_module = module_name in STDLIB_MODULE_NAMES
+        if not (is_builtin_module or is_stdlib_module):
+            filtered_sys_modules.append(module_name)
+
+    deep_modules = set()
+    for module in modules_to_register:
+        all_ast_nodes = []
+
+        if not hasattr(module, "__path__"):
+            if not hasattr(module, "__file__"):
+                continue
+            all_ast_nodes = parse_ast_for_file(
+                python_file_path=getattr(module, "__file__")
+            )
+        else:
+            module_path = getattr(module, "__path__")
+            if len(module_path) == 1:  # Most packaged modules have a source directory
+                python_files = glob.glob(os.path.join(module_path[0], "**.py"))
+                python_files += glob.glob(os.path.join(module_path[0], "**/*.py"))
+            elif len(module_path) == 0:
+                module_file = getattr(module, "__file__")
+                python_files = [module_file]
+            else:
+                raise ValueError("Lug requires source code to get dependencies from module: ", module)
+
+            for python_file_path in python_files:
+                all_ast_nodes += parse_ast_for_file(python_file_path)
+
+        ast_direct_imports = []
+        ast_renamed_imports = []
+        for node in all_ast_nodes:
+            if isinstance(node, ast.Import):
+                ast_direct_imports.append(node)
+            if isinstance(node, ast.ImportFrom):
+                ast_renamed_imports.append(node)
+        for renamed_import in ast_renamed_imports:
+            module_name = renamed_import.module
+            if module_name in filtered_sys_modules and module_name not in modules_to_skip:
+                if "_pytest" in module.__name__ or "url" in module.__name__:
+                    continue
+                deep_modules.add(sys.modules[module_name])
+        for direct_import in ast_direct_imports:
+            module_name = direct_import.names[0].name
+            if module_name in filtered_sys_modules and module_name not in modules_to_skip:
+                if "_pytest" in module.__name__ or "url" in module.__name__:
+                    continue
+                deep_modules.add(sys.modules[module_name])
+    return deep_modules
```

Comment:
The above finds _all_ imports within a dependency, with some limitations. In all tested cases, it causes pickling errors in the cases where it's needed.

That said, there's probably a case we didn't hit – and if by-value cloudpickle is improved, this would open new use-cases. As such, I'd prefer to keep it in place + tested, but not in active use.

Reply:
This part is hard to parse overall, though the variable names and individual lines are straightforward enough to follow. Thoughts on adding 1-line descriptor comments to "sections" of the code within this function, or is that an antipattern?

Reply:
That's a good idea, will add!

Reply:
Added!